### PR TITLE
WsProvider refactoring

### PR DESCRIPTION
**Pull request by kalaninja**

### CODE REVIEW

File: transport/src/main/java/com/strategyobject/substrateclient/transport/ProviderInterface.java [hunk 1]
```diff
@@ -41,6 +41,18 @@
...
+     * Send data to the node
+     *
+     * @param method      The RPC methods to execute
+     * @param params      Encoded parameters as applicable for the method
+     * @param isCacheable Request can be cached
+     * @return future containing result
+     */
+    CompletableFuture<Object> send(String method,
+                                   List<Object> params,
+                                   boolean isCacheable);
```
Comment (vnabiev) [inline transport/src/main/java/com/strategyobject/substrateclient/transport/ProviderInterface.java:54]: The flag doesn't look good to me, I think, using a decorator is more preferable. IMHO, it isn't the responsibility of that interface/method.

File: transport/src/main/java/com/strategyobject/substrateclient/transport/ws/WsProvider.java [hunk 2]
```diff
@@ -167,19 +172,9 @@ public void disconnect() {
...
-                }
-
-                return null;
-            });
-
-        } catch (Exception ex) {
-            logger.error("Error disconnecting", ex);
-            this.emit(ProviderInterfaceEmitted.ERROR, ex);
-            throw ex;
+        val ws = this.webSocket;
```
Comment (vnabiev) [inline transport/src/main/java/com/strategyobject/substrateclient/transport/ws/WsProvider.java:175]: There are some moments that confuse me:
1. The method `connect` is asynchronous, but this isn't.
2. The method doesn't consider whether a connection opened.
3. The next call of `connect` won't have expected effect after `disconnect`.

Comment (vnabiev) [reply transport/src/main/java/com/strategyobject/substrateclient/transport/ws/WsProvider.java:175]: About the third point - they should use the same lock or `AtomicReference` for `whenConnected`.

Comment (vnabiev) [reply transport/src/main/java/com/strategyobject/substrateclient/transport/ws/WsProvider.java:175]: Also I would propose to add the test case for the third point - manual reconnect.

File: transport/src/main/java/com/strategyobject/substrateclient/transport/ProviderInterface.java (hunk 1 above)
Comment (kalaninja) [reply transport/src/main/java/com/strategyobject/substrateclient/transport/ProviderInterface.java]: ok, let it be implemented on the caller side

File: transport/src/main/java/com/strategyobject/substrateclient/transport/ws/WsProvider.java (hunk 2 above)
Comment (kalaninja) [reply transport/src/main/java/com/strategyobject/substrateclient/transport/ws/WsProvider.java:203]: refactored `disconnect` method as well. also added the test.

**Review by vnabiev — APPROVED** [review]

**Pull request merged**